src/treelearner/monotone_constraints.hpp

```diff
@@ -55,14 +94,360 @@ class LeafConstraints {
+      bool use_left_leaf_for_update_right = true;
+      bool use_right_leaf_for_update_left = true;
+      // if the split is on the same feature (categorical variables not supported)
+      // then depending on the threshold,
+      // the current left child may not be contiguous with the original right leaf,
+      // or the current right child may not be contiguous with the original left leaf
+      if (is_split_numerical && inner_feature == split_feature) {
+        if (threshold >= split_threshold) {
+          use_left_leaf_for_update_right = false;
+        }
+        if (threshold <= split_threshold) {
+          use_right_leaf_for_update_left = false;
+        }
+      }
+
+      // go down left
+      if (keep_going_left_right.first) {
+        GoDownToFindLeavesToUpdate(
+            tree, tree->left_child(node_idx),
+            features_of_splits_going_up_from_original_leaf,
+            thresholds_of_splits_going_up_from_original_leaf,
+            was_original_leaf_right_child_of_split, update_max_constraints,
+            split_feature, split_info, use_left_leaf,
+            use_right_leaf_for_update_left && use_right_leaf, split_threshold,
+            best_split_per_leaf);
+      }
+      // go down right
+      if (keep_going_left_right.second) {
+        GoDownToFindLeavesToUpdate(
+            tree, tree->right_child(node_idx),
+            features_of_splits_going_up_from_original_leaf,
+            thresholds_of_splits_going_up_from_original_leaf,
+            was_original_leaf_right_child_of_split, update_max_constraints,
+            split_feature, split_info,
+            use_left_leaf_for_update_right && use_left_leaf, use_right_leaf,
+            split_threshold, best_split_per_leaf);
+      }
+    }
+  }
+
+  std::pair<bool, bool> ShouldKeepGoingLeftRight(
+      const Tree* tree, int node_idx,
+      const std::vector<int>& features_of_splits_going_up_from_original_leaf,
+      const std::vector<uint32_t>&
+          thresholds_of_splits_going_up_from_original_leaf,
+      const std::vector<bool>& was_original_leaf_right_child_of_split) {
+    int inner_feature = tree->split_feature_inner(node_idx);
+    uint32_t threshold = tree->threshold_in_bin(node_idx);
+    bool is_split_numerical = tree->IsNumericalSplit(node_idx);
+
+    bool keep_going_right = true;
+    bool keep_going_left = true;
+    // left and right nodes are checked to find out if they are contiguous with
+    // the original leaves if so the algorithm should keep going down these nodes
+    // to update constraints
+    if (is_split_numerical) {
+      for (size_t i = 0;
+           i < features_of_splits_going_up_from_original_leaf.size(); ++i) {
+        if (features_of_splits_going_up_from_original_leaf[i] ==
+            inner_feature) {
+          if (threshold >=
+                  thresholds_of_splits_going_up_from_original_leaf[i] &&
+              !was_original_leaf_right_child_of_split[i]) {
+            keep_going_right = false;
```

Comment:
Can we speedup here (and in the next `if`) by not running the rest of the loop when the opposite direction is already `false` by something like
```
keep_going_right = false;
if (!keep_going_left) {
  break;
}
```
and will it make any sense?

Reply:
Yes that make sense, that should speed up the function. I just pushed the change, @StrikerRUS let me know if there is anything else.